plugins/inputs/loggregator_rlp/envelope_writer.go

```diff
@@ -0,0 +1,249 @@
+	"code.cloudfoundry.org/loggregator-agent/pkg/diodes"
+	"github.com/influxdata/telegraf"
+)
+
+var (
+	validMetricNameExpression = regexp.MustCompile(`^[a-zA-Z_:\-.][a-zA-Z0-9_:\-.]*$`)
+	validLabelExpression      = regexp.MustCompile(`^[a-zA-Z_\-.][a-zA-Z0-9_\-.]*$`)
+)
+
+const (
+	droppedMetricName              = "telegraf_dropped_envelopes"
+	ingressMetricName              = "telegraf_ingress_envelopes"
+	egressMetricName               = "telegraf_egress_envelopes"
+	droppedDeltaCountersMetricName = "telegraf_dropped_delta_counters"
+)
+
+type EnvelopeWriter struct {
+	acc            telegraf.Accumulator
+	stop           chan struct{}
+	wg             sync.WaitGroup
+	envelopesQueue *diodes.OneToOneEnvelopeV2
+
+	ingressedEnvelopes   *uint64
+	egressedEnvelopes    *uint64
+	droppedEnvelopes     *uint64
+	droppedDeltaCounters *uint64
+}
+
+func NewEnvelopeWriter(acc telegraf.Accumulator, internalMetricsInterval time.Duration) *EnvelopeWriter {
+	w := &EnvelopeWriter{
+		acc:  acc,
+		stop: make(chan struct{}),
+
+		ingressedEnvelopes:   new(uint64),
+		egressedEnvelopes:    new(uint64),
+		droppedEnvelopes:     new(uint64),
+		droppedDeltaCounters: new(uint64),
+	}
+
+	w.envelopesQueue = diodes.NewOneToOneEnvelopeV2(10000, gendiodes.AlertFunc(func(missed int) {
+		w.incrementMetric(w.droppedEnvelopes, uint64(missed))
+	}))
+
+	go w.handleEnvelopes()
+	go w.periodicallyReportInternalMetrics(internalMetricsInterval)
+
+	return w
+}
+
+func (w *EnvelopeWriter) periodicallyReportInternalMetrics(interval time.Duration) {
+	w.reportInternalMetrics()
+
+	intervalTick := time.Tick(interval)
+	for {
+		select {
+		case <-w.stop:
+			return
+		case <-intervalTick:
+			w.wg.Add(1)
+			w.reportInternalMetrics()
+			w.wg.Done()
+		}
+	}
+}
+
+func (w *EnvelopeWriter) Write(env *loggregator_v2.Envelope) {
+	w.envelopesQueue.Set(env)
+	w.incrementMetric(w.ingressedEnvelopes, 1)
+}
+
+func (w *EnvelopeWriter) handleEnvelopes() {
+	for {
+		select {
+		case <-w.stop:
+			return
+		default:
+			e, ok := w.envelopesQueue.TryNext()
+			if !ok {
+				time.Sleep(10 * time.Millisecond)
+				continue
+			}
+			w.wg.Add(1)
+			w.handleEnvelope(e)
+			w.wg.Done()
+		}
+	}
+}
+
+func (w *EnvelopeWriter) handleEnvelope(env *loggregator_v2.Envelope) {
+	switch env.GetMessage().(type) {
+	case *loggregator_v2.Envelope_Counter:
+		w.writeCounter(env)
+	case *loggregator_v2.Envelope_Gauge:
+		w.writeGauge(env)
+	case *loggregator_v2.Envelope_Timer:
+		w.writeTimer(env)
+	default:
+		return
+	}
+
+	w.incrementMetric(w.egressedEnvelopes, 1)
+}
+
+func (w *EnvelopeWriter) writeCounter(env *loggregator_v2.Envelope) {
+	ts := time.Unix(0, env.GetTimestamp())
+	counter := env.GetCounter()
+	if !validName(counter.GetName()) {
+		return
+	}
+
+	if counter.GetTotal() == 0 && counter.GetDelta() > 0 {
+		w.incrementMetric(w.droppedDeltaCounters, 1)
+		return
+	}
+
+	fields := map[string]interface{}{
+		"counter": counter.GetTotal(),
```

Comment:
The name of this field indicates that we may not be using the measurement and field names using the current Telegraf style.  Most likely we would want to set the field name to `counter.GetName()` and the measurement would be the same across all metrics, perhaps `loggregator_rlp`.  Add some example data and we can take a closer look.